### PR TITLE
[SPARK-15538][SPARK-15536][SQL] Truncate table fixes round 2

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

Two more changes:
(1) Fix truncate table for data source tables (only for cases without `PARTITION`)
(2) Disallow truncating external tables or views
## How was this patch tested?

`DDLSuite`

### CODE REVIEW

**Comment by andrewor14** [comment]
@hvanhovell @yhuai

**Comment by sureshthalamati** [comment]
@andrewor14 
I tried  this PR.  It  does not seem to work/raise error if user attempts to truncate on partition table  with out specifying the partition spec.
val df = Seq((1 , "john", "CA") ,(2,"Mike", "NY"), (3, "Robert", "CA")).toDF("id", "name", "state")
df.write.format("parquet").partitionBy("state").saveAsTable("emp16")
sql("truncate table emp16") 
sql("select count(*) from emp16").show

**Comment by lianhuiwang** [comment]
@andrewor14 I think catalog.invalidateTable(tableName) can fix @sureshthalamati 's problem.

**Comment by andrewor14** [comment]
@sureshthalamati what behavior did you expect? I would think that truncating a partitioned table without specifying the specs should just delete data from all partitions:
```
scala> sql("select * from emp16").show()
+---+------+-----+
| id|  name|state|
+---+------+-----+
|  3|Robert|   CA|
|  1|  john|   CA|
|  2|  Mike|   NY|
+---+------+-----+
scala> sql("TRUNCATE TABLE emp16")
scala> sql("select * from emp16").show()
+---+----+-----+
| id|name|state|
+---+----+-----+
+---+----+-----+
```

**Comment by yhuai** [comment]
{{sql("truncate table emp16") }} should delete all partitions, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 1]
```diff
@@ -289,37 +289,53 @@ case class TruncateTableCommand(
...
+        s"Operation not allowed: TRUNCATE TABLE ... PARTITION is not supported " +
+        s"for tables created using the data sources API: '$tableName'")
+    }
+    val locations = if (partitionSpec.isDefined) {
+      catalog.listPartitions(tableName, partitionSpec).map(_.storage.locationUri)
     } else {
-      val locations = if (partitionSpec.isDefined) {
-        catalog.listPartitions(tableName, partitionSpec).map(_.storage.locationUri)
+      if (table.partitionColumnNames.nonEmpty) {
+        catalog.listPartitions(tableName).map(_.storage.locationUri)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: This is for hive table, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 2]
```diff
@@ -289,37 +289,53 @@ case class TruncateTableCommand(
...
+          case NonFatal(e) =>
+            throw new AnalysisException(
+              s"Failed to truncate table '$tableName' when removing data of the path: $path " +
+                s"because of ${e.toString}")
         }
       }
     }
+    // After deleting the data, invalidate the table to make sure we don't keep around a stale
+    // file relation in the metastore cache.
+    sparkSession.sessionState.invalidateTable(tableName.unquotedString)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Seems we also need to drop the cached data in the InMemoryColumnarStore?

**Comment by sureshthalamati** [comment]
@andrewor14  The behavior you mentioned is the one I was expecting also.  For some reason it does not work in my env. I was trying for a spark-shell .  What catalog are you using ?  Probably I just need to try with a clean build.

**Comment by sureshthalamati** [comment]
@andrewor14  After I cleaned up the old  metastore_db ,  truncate worked as expected.  Thanks

**Comment by sureshthalamati** [comment]
@andrewor14  one minor thing I noticed when I looked at the table data directories;  In the case of regular tables truncate keeps the partition directories, for data source table partition directories also deleted.  It may be ok , thought I will mention it incase if we want to be consistent

**Comment by andrewor14** [comment]
I tested partitioned data source tables too. If you add data back then the partitions will be created again. I think that's OK.

**Comment by yhuai** [comment]
lgtm

**Comment by sureshthalamati** [comment]
lgtm

**Comment by hvanhovell** [comment]
LGTM

**Comment by andrewor14** [comment]
Thanks for all the LGTMs. I'm going to merge this into master 2.0.

**Pull request closed without merging**